[MRG] Implement joint_plot

jona-sassenhagen opened the pull request:
Add a viz function for a joint topo and butterfly plot, EEGLAB style, see #2664

Try it with ```mne.viz.evoked.joint_plot(evoked)```.

Edit: now also with a method, `evoked.plot_joint()`

![unknown-50](https://cloud.githubusercontent.com/assets/4321826/11963554/1d99c45c-a8e8-11e5-85d3-d40ec77cbb4d.png)

jona-sassenhagen commented:
Can I get some feedback? @jaeilepp @kingjr @agramfort ..?

kingjr commented:
That looks good. I would use continuous gray lines to minimize the number
of edge (and this visual complexity) increase the topo size (the butterfly
time course is generally mainly readable in terms of GDP although with your
new coloring style you can be a bit more informative) and add a white
background to the title so that the lines don't strike the text.


On Tuesday, 22 December 2015, jona-sassenhagen <notifications@github.com>
wrote:

> Can I get some feedback? @jaeilepp <https://github.com/jaeilepp> @kingjr
> <https://github.com/kingjr> @agramfort <https://github.com/agramfort> ..?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/2738#issuecomment-166724291>
> .
>

jaeilepp commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,142 @@ def plot_snr_estimate(evoked, inv, show=True):
+        return figs
+
+    import matplotlib.pyplot as plt
+    f = plt.figure()
```
f -> fig

jaeilepp commented:
The figure is a bit off for me:
![screenshot_2015-12-23_10-31-11](https://cloud.githubusercontent.com/assets/3456414/11974020/5e488fce-a960-11e5-98e9-2db067e47a8d.png)

The picks argument doesn't seem to work for me.

jaeilepp commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,142 @@ def plot_snr_estimate(evoked, inv, show=True):
+    if len(ch_types) > 1:
+        figs = list()
+        for t in ch_types:  # pick only the corresponding channel type
+            e = evoked.copy().pick_channels([info["ch_names"][idx]
```
Avoid single letter variables.

jona-sassenhagen commented:
@jaeilepp can you show me the command you tried?

jaeilepp commented:
```
picks = mne.pick_types(evoked.info, meg='mag')
joint_plot(evoked, picks=picks)
```
Anything with picks fails for me.

jona-sassenhagen commented:
And you didn't use the `times` kwarg, this is with auto picks? Maybe I did something wrong with sampling rate ...

jaeilepp commented:
The problem seems to be that ``evoked.pick_channels`` only accepts channel names and by passing indices as picks it fails.

jaeilepp commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,142 @@ def plot_snr_estimate(evoked, inv, show=True):
+        exclude = [ch for ch in info['bads'] if ch in info['ch_names']]
+    evoked = evoked.copy().drop_channels(exclude)
+    if picks is not None:
+        evoked = evoked.pick_channels(picks)
```
By removing this check it seems to work.

jaeilepp replied:
Actually... It works, but shows all the channels. So it seems that picks argument is ignored.

jona-sassenhagen replied:
I guess it only removes the intersection of the channels it's provided with, and the channels in the data set. Will fix.

jona-sassenhagen commented:
Oh well :) will fix. But what about the times, that seems strange. You're not setting `times`?

jaeilepp commented:
No, I'm not setting times. I'm not sure what you mean by it.

jona-sassenhagen commented:
Ah, sorry, my fault, I misunderstood the plot.

You can set the times to show topomaps at with `topomap_args={times=[0., .2, .35]}` or so.

jona-sassenhagen commented:
@kingjr I think I'll just remove the title string (or join it with the figure title).

jaeilepp commented:
I see... I wonder if we could avoid dicts to pass kwargs to topomap. I know matplotlib uses a lot of hidden kwargs and it's really annoying.
I would just condense the amount of parameters to the minimum and expose them to the user as optional params. Otherwise I'm afraid the users never bother going through the docs to find out all the possible options.

jona-sassenhagen commented:
I'm considering that. But my idea would be to instead have good defaults so you rarely have to touch them.
The thing is, I don't want to support only a subset of plot_evoked and plot_topomap params, but the full joint set is a LOT of params.

Using kwarg dicts is the seaborn way of doing things.

But I agree, it's a problematic topic.

jona-sassenhagen commented:
http://nbviewer.ipython.org/gist/jona-sassenhagen/1533861e529110aa2fbc
Here is an example gist.

jona-sassenhagen commented:
Note: EEGLAB also has these for Time-Frequency plots, power spectra, and even ERPimages. But this is just for Butterfly ERPs for now.

![425px-channelspectra](https://cloud.githubusercontent.com/assets/4321826/11994752/6c55de20-aa44-11e5-8185-b92b5ccbe75b.gif)

jona-sassenhagen commented:
I've addressed the comments.

@jaeilepp I think I got everything you had noted.

@agramfort what do you think of the kwarg dict call signature? See gist: http://nbviewer.ipython.org/gist/jona-sassenhagen/1533861e529110aa2fbc
It's Seaborn's way of exposing the full arg set of two joint functions.

@kingjr IMO straight connection lines don't look as good. They actually feel *more* salient to me.

![screen shot 2016-01-08 at 13 57 19](https://cloud.githubusercontent.com/assets/4321826/12198427/06b282b2-b610-11e5-8ed9-a45c78487872.png)
![screen shot 2016-01-08 at 13 56 59](https://cloud.githubusercontent.com/assets/4321826/12198428/06d19f6c-b610-11e5-8bc7-ba2a694e93d5.png)

kingjr commented:
I prefer continuous but it's your call, it's a detail. Also, I would put the line behind the butterfly (or is it the case already? there seems to be some kind of a transparency?)

jona-sassenhagen commented:
No, you're right, I had accidentally only set the z order for half of the lines.
The gray connection lines are indeed slightly transparent, but they'll all be set to the lowest z order.

Any other opinions on continuous/dashed lines?

Maybe this is just my very subjective preference - they're dashed in EEGLAB ...

jaeilepp commented on mne/evoked.py:
```diff
@@ -18,8 +18,8 @@
 from .viz import (plot_evoked, plot_evoked_topomap, plot_evoked_field,
-                  plot_evoked_image, plot_evoked_topo)
-from .viz.evoked import _plot_evoked_white
+                  plot_evoked_image, plot_evoked_topo, )
```
Remove ',' at the end

jaeilepp commented on mne/evoked.py:
```diff
@@ -727,6 +727,47 @@ def plot_white(self, noise_cov, show=True):
+        .. versionadded:: 0.12.0
+        """
+        return joint_plot(evoked, title=title, picks=picks, exclude=exclude,
+                          show=True, ts_args=ts_args,
```
show=show

jaeilepp commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,152 @@ def plot_snr_estimate(evoked, inv, show=True):
+                                   topomap_args=topomap_args))
+        return figs
+
+    import matplotlib.pyplot as plt
```
Maybe this should go to the beginning of the function.

jaeilepp commented:
Maybe I would expose ``times`` param to ``joint_plot`` separately since it feels more like a feature of ``joint_plot``. In addition you could have ``topomap_args`` where you could set rest of the parameters for topomap. Wdyt?

jaeilepp commented:
Otherwise LGTM.

jona-sassenhagen commented:
>Maybe I would expose times param to joint_plot separately since it feels more like a feature of joint_plot.

I actually had it like that originally - under the hood, it would be more naturally.

Do interactive plots look good for you now?

jaeilepp commented:
I'm not sure what you mean by interactive, but it renders much better now. Maybe I would lift the title a little bit. With ``y=98`` it looks better to my eye.

jona-sassenhagen commented:
Okay will do.

jaeilepp commented:
Just noticed that with weird picks the sensor positions are not correct. You can try ``picks = mne.pick_types(evoked.info, meg='grad')[:10]`` with sample dataset. The same problem exists with ``spatial_colors`` so maybe we'll tackle it in different PR. I'll open an issue

jona-sassenhagen commented:
Good. I'll push the latest fixed (including continuous connection lines, unless you think dashed looks better) and set to MRG.

jaeilepp commented:
If the title is lifted, there is more room for the topomaps to be bigger. I think in there is room in the horizontal dimension too, since no one is expected to draw more than three topomaps. Right?

jaeilepp commented:
Or is it somehow dynamically sized? Anyway, I think the topomaps should be a little bit larger.

jona-sassenhagen commented:
I sometimes plot a lot of topomaps, like 7. But yes, Jean-Remi also wanted the topoplots bigger, I'll see what I can do.

jona-sassenhagen commented:
Okay, Travis should come back green now and all comments are addressed.

jona-sassenhagen commented:
Sorry I didn't get this done in time for HBM :(

kingjr commented:
According to my source, It's still open until 3pm

On 8 January 2016 at 12:13, jona-sassenhagen <notifications@github.com>
wrote:

> Sorry I didn't get this done in time for HBM :(
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/2738#issuecomment-170060836>
> .
>

jona-sassenhagen commented:
Yeah but I doubt Alex wants to first merge this, and then redo the submission today ...

Eric89GXL commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,158 @@ def plot_snr_estimate(evoked, inv, show=True):
+
+
+def joint_plot(evoked, times="peaks", title='', picks=None, exclude=list(),
+               show=True, ts_args=dict(), topomap_args=dict()):
```
do not use mutable default arguments. So `exclude=(), ts_args=None, topomap_args=None` and later have e.g. `topomap_args = dict() if topomap_args is None else topomap_args`

jona-sassenhagen replied:
I think I got something wrong here - I thought this would be a problem with `arg=[]`, but not with `arg=list()`?

Eric89GXL replied:
```
>>> def mutey(a=list()):
>>>     a.append('foo')
>>>     print(a)
>>> mutey()
['foo']
>>> mutey()
['foo', 'foo']
```

Eric89GXL commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,158 @@ def plot_snr_estimate(evoked, inv, show=True):
+                    if channel_type(info, idx) in data_types])
+    # if multiple sensor types: one plot per channel type, recursive call
+    if len(ch_types) > 1:
+        print(ch_types)
```
from debug

jona-sassenhagen replied:
Oh god yes. I feel like you caught me without my pants on.

Eric89GXL replied:
I actually do this all the time, too :) I catch myself at least some of those times by reading through my PR diff myself one more time before asking for MRG reviews, but I don't even catch all of them that way

Eric89GXL commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,158 @@ def plot_snr_estimate(evoked, inv, show=True):
+        elif times == "auto":
+            times = np.linspace(evoked.times[0], evoked.times[-1], 5)
+    elif np.isscalar(times):
+        times = [times]
```
is this duplicated from somewhere? If so, please refactor into e.g. `times = _process_times(evoked, topomap_args.get("peaks", 3)` to make DRY

jona-sassenhagen replied:
It's a slight adaption of the default for `plot_topomap`. There, the default is 7 peaks or 10 regular plots for auto times. But I think these joint plots get much too messy with 7 and 10, so I'm using 3 and 5 respectively.

Eric89GXL replied:
That's fine, make the DRY function take the default levels as parameters, then.
```
def _process_times(times, inst, peaks=7, auto=10):
   ...
```
Then here you do:
```
times = _process_times(times, evoked, peaks=3, auto=5)
```
make sense?

jona-sassenhagen replied:
Hm ... that would mean updating `plot_topomap` right?

And you mean a function that takes either nothing but an evoked, or a number of floats, or a string (and possibly a number) for the peak/auto options?

(I'd think we want a function that always returns a series of floats, and one that only computes peaks if it's asked to, because computing peaks might become expensive.)

Eric89GXL replied:
Yeah you'd update every place that uses the same logic. I'm not familiar enough with the viz code to know, but you can probably find out by doing `git grep _find_peaks`. `plot_topomap` probably uses it, too, yeah.

Here is the filled-in function:
```
def _process_times(times, inst, peaks_val=7, auto_val=10):
    if isinstance(times, string_types):
        if times == "peaks":
            times = _find_peaks(evoked, topomap_args.get("peaks", peaks_val))
        elif times == "auto":
            times = np.linspace(evoked.times[0], evoked.times[-1], auto_val)
    elif np.isscalar(times):
        times = [times]
    return times
```
Make sense?

jona-sassenhagen replied:
Ah I was a bit confused by the example call of the function using both peaks and auto, but I get why you did that now. This probably belongs in `mne.evoked`, since it only applies to evoked objects, but not exclusively to viz?

jona-sassenhagen replied:
Also I'm pretty sure you *wrote* the part in `plot_topomap` I'd be changing here (the peak finder).

Eric89GXL replied:
> This probably belongs in mne.evoked, since it only applies to evoked objects, but not exclusively to viz?

Is it only used by viz functions? By evoked viz functions? If so I'd put it in the mne/viz/evoked.py

> Also I'm pretty sure you wrote the part in plot_topomap I'd be changing here (the peak finder).

That's certainly possible, I don't remember. There presumably wasn't a need to have this logic go in its own function before since (I guess) it was only used in one place, but now it makes sense to separate it out to keep the code DRY.

jona-sassenhagen replied:
>Is it only used by viz functions? By evoked viz functions? If so I'd put it in the mne/viz/evoked.py

Right now yes, but I think in principle finding the largest n peaks would be useful for all kinds of functions. But okay, I guess the peak finder method is too ad hoc to be useful for rigorous statistics. I'll put it under viz.

Eric89GXL replied:
And keep in mind that this `_process_times` function is really pretty specific to triaging the behavior of a specific keyword argument, even if the functions it calls (peak finding or whatever) might end up being used elsewhere.

Eric89GXL commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,158 @@ def plot_snr_estimate(evoked, inv, show=True):
+                  linestyle='-', linewidth=1.5, alpha=.66, zorder=-1)
+
+
+def joint_plot(evoked, times="peaks", title='', picks=None, exclude=list(),
```
let's make this function private to push people to use `evoked.plot_joint`

Eric89GXL commented on mne/evoked.py:
```diff
@@ -727,6 +727,47 @@ def plot_white(self, noise_cov, show=True):
                                   rank=None, show=show)
 
+    def plot_joint(evoked, title='', picks=None, exclude=list(), show=True,
+                   ts_args=dict(spatial_colors=True), topomap_args=dict()):
```
same thing about mutable defaults here. In the docstring you can say that `ts_args=None` will be translated to `dict(spatial_colors=True)`

jona-sassenhagen commented:
Travis should come back green ...

jona-sassenhagen commented:
Done! Any other comments?

teonlamont commented:
naive question but why are the field maps of the topoplot outside the head?

jona-sassenhagen commented:
Because I'm defaulting to 'skirt' for the outlines. If you don't want that, you can pass 'outlines'='head' in the topomap_args.

agramfort commented on mne/evoked.py:
```diff
@@ -727,6 +727,57 @@ def plot_white(self, noise_cov, show=True):
+        Parameters
+        ----------
+        evoked : instance of Evoked
+            The evoked instance.
```
remove for the method

agramfort commented on mne/viz/evoked.py:
```diff
@@ -385,7 +385,8 @@ def _plot_evoked(evoked, picks, exclude, unit, show,
     fig.canvas.draw()  # for axes plots update axes.
-    tight_layout(fig=fig)
+    if set_tight_layout:
+        tight_layout(fig=fig)
```
does this have an impact on any of the plots in the gallery?

jona-sassenhagen replied:
No, it defaults to true.

agramfort commented on mne/viz/evoked.py:
```diff
@@ -393,7 +394,8 @@ def _plot_evoked(evoked, picks, exclude, unit, show,
                 scalings=None, titles=None, axes=None, gfp=False,
-                window_title=None, spatial_colors=False):
+                window_title=None, spatial_colors=False,
+                set_tight_layout=True):
```
do we really need this extra arg? we return the fig so people can do plot adjustments without making our API more complex

jona-sassenhagen replied:
It crashes for me with tight layout on.

jaeilepp replied:
How about only exposing this param for the private function below and using it in the code for ``joint_plot``? I don't think the public function needs this.

agramfort replied:
+1 I don't like this set_tight_layout param.

we've always managed to handle it in the past across mpl backends

agramfort commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,164 @@ def plot_snr_estimate(evoked, inv, show=True):
+    data_types = ['eeg', 'grad', 'mag', 'seeg']
+    ch_types = set([channel_type(info, idx)
+                    for idx in range(info['nchan'])
+                    if channel_type(info, idx) in data_types])
```
ch_types = set(ch_type for ch_type in ['eeg', 'grad', 'mag', 'seeg'] if ch_type in evoked)

should do it.

agramfort commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,164 @@ def plot_snr_estimate(evoked, inv, show=True):
+            ev_ = evoked.pick_channels([info["ch_names"][idx]
+                                        for idx in range(info['nchan'])
+                                        if channel_type(info, idx) == t],
+                                       copy=True)
```
ev.pick_types should be used here.

jona-sassenhagen replied:
I can't figure out how to use the pick_types syntax though. How do I convert a string into a kwarg? E.g., how do I translate `'eeg'` into `pick_types(eeg=True, meg=False ...)`? Do we have an example somewhere?

jaeilepp replied:
``topomap`` does it like this:
```
        if ch_type == 'eeg':
            picks = pick_types(info, meg=False, eeg=True, ref_meg=False,
                               exclude='bads')
        else:
            picks = pick_types(info, meg=ch_type, ref_meg=False,
                               exclude='bads')
```

jona-sassenhagen replied:
6-7 lines to my 4, and doesn't work with seeg ... is this really an improvement?

agramfort replied:
let's add a private function in pick.py to do this. We can already factorize the code in 2 locations.

jona-sassenhagen replied:
Add the function in this PR?

jona-sassenhagen replied:
I'd prefer not to and rather refactor later because picks confuse me ...

agramfort commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,164 @@ def plot_snr_estimate(evoked, inv, show=True):
+    # butterfly/time series plot
+    ts_ax = fig.add_subplot(212)
+    ts_args_pass = dict((k, v) for k, v in ts_args.items()
+                        if k not in ["axes", "show", "colorbar"])
```
dict comprehension does not work with python 2.6 we still support for now

jona-sassenhagen replied:
IDGI ... Isn't it this syntax that doesn't work in py2.6:
`{k:v for k, v in iterable}`

As-is, this line passes Travis. And we're using this syntax in other places.

jona-sassenhagen replied:
I just tried, and this works on py2.6:

```
Python 2.6.6 (r266:84292, Aug 12 2014, 07:57:07)
[GCC 4.4.5] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> dict((k, k) for k in range(3))
{0: 0, 1: 1, 2: 2}
```

And here is an example of the above syntax from MNE code where `{k:v for k, v in iterable}` syntax was exchanged for the `dict((k, v) for k, v in iterable)` syntax for the sake of py2.6 compatibility:

https://github.com/mne-tools/mne-python/commit/368b0c085b359484f84e3074209125f02c60c3da

Or am I missing your point?

agramfort commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,164 @@ def plot_snr_estimate(evoked, inv, show=True):
+    ts_args_pass = dict((k, v) for k, v in ts_args.items()
+                        if k not in ["axes", "show", "colorbar"])
+    ts_args_pass.update(dict(spatial_colors=ts_args.get("spatial_colors",
+                                                        True)))
```
ts_args_pass["spatial_colors"] = True if "spatial_colors" not in ts_args_pass

is more readable

agramfort commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,164 @@ def plot_snr_estimate(evoked, inv, show=True):
+    # slightly convoluted due to colorbar placement and for vertical alignment
+    ts = len(times) + 2
+    map_ax = [plt.subplot(4, ts, x + 2 + ts) for x in range(ts - 2)]
+    cbar_ax = plt.subplot(4, 3 * (ts + 1), 6 * (ts + 1))
```
you should adjust the figure grid as currently there is a lot of blank space between suptitle and the topos.

see for example how it's done in plot_epochs_image

jona-sassenhagen replied:
The problem is, I can't set the title in a way so that it's guaranteed to stay at a certain distance to the upper plots. This means if I set it too low, for narrow plots, it's not guaranteed to not overlap with the topoplots.

In `plot_epochs_image*, this is not a problem because we can use the axis title. But that doesn't work here, because the title of the top axes is already in use.

Maybe @jaeilepp has an idea?

jona-sassenhagen replied:
![screen shot 2016-01-10 at 13 15 10](https://cloud.githubusercontent.com/assets/4321826/12221504/fed05e86-b79d-11e5-8167-e51bc7093def.png)

This is how plots look right now.

jona-sassenhagen replied:
Okay, I found something that works with the scaling. It's a bit weird, but it works.

jaeilepp replied:
I'm not sure how I've done it before, but sometimes ``ax.text`` works better than actual title items.

jona-sassenhagen replied:
Yeah, that's how I'm doing it now. Works better.

agramfort commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,164 @@ def plot_snr_estimate(evoked, inv, show=True):
+
+    # topomap
+    topomap_args_pass = dict((k, v) for k, v in topomap_args.items() if
+                             k not in ["times", "axes", "show", "colorbar"])
```
dict comprehension

agramfort commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,164 @@ def plot_snr_estimate(evoked, inv, show=True):
+    topomap_args_pass = dict((k, v) for k, v in topomap_args.items() if
+                             k not in ["times", "axes", "show", "colorbar"])
+    topomap_args_pass.update(dict(outlines=topomap_args.get("outlines",
+                                                            "skirt")))
```
same rmk

agramfort commented on mne/viz/tests/test_topo.py:
```diff
@@ -77,6 +77,11 @@ def test_plot_topo():
+    # Test jointplot
+    evoked.plot_joint()
+    evoked.plot_joint(title='test', ts_args=dict(spatial_colors=True),
+                      topomap_args=dict(colorbar=True, times=[0.]))
```
you should update the plot_evoked_erf_erp.py example to show case this new feature

jona-sassenhagen replied:
Then we're returning 4 figures in that function ... a lot IMO.
I'd prefer to use an example with only one channel type. I want to make an example for linear_regression_raw with continuous covariates anyways because it's not quite clear to people how it works for now, I just need to make a file for that available.

But I guess for now it'd be best to use plot_evoked_erf_erp.

jona-sassenhagen replied:
I've added a new example. It bothered me that plot_evoked_erf_erp returned 4 figures, and I also thought the dict arguments needed a dedicated example.

agramfort commented:
that's it for me now.

nice work @jona-sassenhagen

jaeilepp commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,172 @@ def plot_snr_estimate(evoked, inv, show=True):
+    topomap_args_pass = dict((k, v) for k, v in topomap_args.items() if
+                             k not in ["times", "axes", "show", "colorbar"])
+    topomap_args_pass["outlines"] = (topomap_args["outlines"] if "outlines"
+                                     in topomap_args else 'skirt')
```
You could do ``topomap_args_pass['outlines'] = topomap_args.get('outlines', 'skirt')``. And for consistency, use single quotes when possible.

jona-sassenhagen replied:
Alex just told me to do it like this, I had the `.get` one before :)

agramfort commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,172 @@ def plot_snr_estimate(evoked, inv, show=True):
+        cbar.locator = ticker.MaxNLocator(nbins=5)
+        cbar.update_ticks()
+
+    plt.subplots_adjust(left=.1, right=.9, bottom=.13,
```
right=.94 is nicer

jona-sassenhagen replied:
For default sized plots yes, but then it starts swallowing the colorbar tick labels if you resize the window.

agramfort replied:
>
> For default sized plots yes, but then it starts swallowing the colorbar tick labels if you resize the window.
>
>
not for me.

jona-sassenhagen replied:
![screen shot 2016-01-11 at 14 33 20](https://cloud.githubusercontent.com/assets/4321826/12234883/5ba7d486-b870-11e5-8a0f-b4984c7635d1.png)

This is what mags look like for me if I decrease size by approx. half and right=.94.

Admittedly, I should probably leave more borders at bottom and left ...

agramfort replied:
ok maybe 0.94 is a bit too much.

but to me you should not have to change the figsize manually as it will
always break rendering.

jona-sassenhagen replied:
Haha I've spend so much time trying to make the function work decently with manual resizing because I thought that's an expected use case (I'd never do it personally) ... Okay.

agramfort commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,172 @@ def plot_snr_estimate(evoked, inv, show=True):
+                                    topomap_args=topomap_args))
+        return figs
+
+    fig = plt.figure()
```
fig = plt.figure(figsize=(8.0, 4.2))

jona-sassenhagen commented:
@agramfort , is it okay if refactoring picking types is set for another time, and I just adjust plot size/border defaults and then it's merged?

jaeilepp commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,172 @@ def plot_snr_estimate(evoked, inv, show=True):
+        finds time points automatically by checking for 3 local maxima in
+        Global Field Power.
+    title : str | None
+        The title. If `None`, supress printing channel type.
```
I think the correct way is to use double backward ticks. And add ``defaults to ''``.

jaeilepp commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,172 @@ def plot_snr_estimate(evoked, inv, show=True):
+        A dict of `kwargs` that are forwarded to `evoked.plot_topomap`
+        to style the topomaps. `axes` and `show` are ignored. If `times`
+        is not in this dict, automatic peak detection is used. Beyond that,
+        if `None`, no customizable arguments will be passed.
```
Double ticks here too.

jaeilepp commented on mne/viz/evoked.py:
```diff
@@ -875,3 +880,172 @@ def plot_snr_estimate(evoked, inv, show=True):
+    evoked : instance of Evoked
+        The evoked instance.
+    times : float | array of floats | "auto" | "peaks".
+        The time point(s) to plot. If "auto", 5 evently spaced topographies
```
evently -> evenly. And please add 'defaults to' for all params.

jona-sassenhagen commented:
Comments addressed, Travis is good. 

Nice, coverage is up.

jaeilepp commented:
+1 for merge

agramfort commented:
great merging !

agramfort commented:
thanks @jona-sassenhagen !

jona-sassenhagen commented:
Thanks for reviews everyone.

jona-sassenhagen commented:
@agramfort  can I make a PR for whats_new? I think I forgot that for a few contribs of mine ...

agramfort commented:
@jona-sassenhagen can just send a PR to update what's new?

jona-sassenhagen commented:
Yes, that's what I meant. Will do.
